File: fsharp/strain/Strain.fs
module Seq

    let keep (pred:'a -> bool) (xs:'a seq):'a seq =
        seq{
            for x in xs do
                if x |> pred then
                    yield x}

    let discard (pred:'a -> bool) (xs:'a seq):'a seq =
        seq{
            for x in xs do
                if x |> pred |> not then
                    yield x}

File: solutions/scala/strain/2/src/main/scala/Strain.scala
object Strain
{
    def keep[A](xs:Seq[A], pred:A => Boolean):Seq[A] =
    {
        for (x <- xs if pred(x))
            yield x
    }

    def discard[A](xs:Seq[A], pred:A => Boolean):Seq[A] =
    {
        for (x <- xs if !pred(x))
            yield x
    }
}